### PR TITLE
Add Pamphlets

### PR DESCRIPTION
**Pull request by gamma-delta**
Pamphlets are a new type of book that have exactly one category. Opening the book brings you to a special landing page that shows the list of entries in that category, instead of a list of categories. This is intended for mods like Retro Candles which need more information than you can cleanly put on tooltips, but not so much information that it would be fruitful to have multiple categories.

To make a book into a pamphlet, just set `"pamphlet_category": "bookid:categoryid"` in the book.json.

Includes documentation!

Also while I was at I impled that feature I've been wanting for a while where a bare string (instead of an object) in a page table automagically becomes a `patchouli:text` page!

### CODE REVIEW

File: Common/src/main/java/vazkii/patchouli/client/book/BookContentsBuilder.java [hunk 1]
```diff
@@ -99,11 +100,28 @@ public BookContents build(Book book) {
 			}
 		});
 
+		BookCategory pamphletCategory;
+		if (book.pamphletCategoryID == null) {
+			pamphletCategory = null;
+		} else {
```
Comment (williewillus) [inline Common/src/main/java/vazkii/patchouli/client/book/BookContentsBuilder.java]:
```suggestion
		BookCategory pamphletCategory = null;
		if (book.pamphletCategoryID != null) {
```

File: Common/src/main/java/vazkii/patchouli/client/book/ClientBookRegistry.java [hunk 2]
```diff
@@ -97,6 +97,13 @@ public static class LexiconPageAdapter implements JsonDeserializer<BookPage> {
 
 		@Override
 		public BookPage deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
+			if (json instanceof JsonPrimitive prim && prim.isString()) {
+				// Shortcut: make strings instead of objects shortcut to text pages
+				var out = new PageText();
+				out.setText(prim.getAsString());
+				return out;
+			}
+
```
Comment (williewillus) [inline Common/src/main/java/vazkii/patchouli/client/book/ClientBookRegistry.java]: Please separate unrelated changes out into a separate diff/PR.

File: Common/src/main/java/vazkii/patchouli/client/book/gui/GuiBookPamphletLanding.java [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+import vazkii.patchouli.common.book.Book;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static vazkii.patchouli.client.book.gui.GuiBookEntryList.ENTRIES_IN_FIRST_PAGE;
+import static vazkii.patchouli.client.book.gui.GuiBookEntryList.ENTRIES_PER_PAGE;
+
+// A lot of this is a copy of GuiBookLanding
```
Comment (williewillus) [inline Common/src/main/java/vazkii/patchouli/client/book/gui/GuiBookPamphletLanding.java]: Hmm. I don't like all this copypasting. What about this?:

Modify GuiBookLanding to directly check if the book only has one category. If so, display the entries of that category inline on the right, else do the existing logic. You can probably factor out the entry button building code from the existing place so you don't duplicate that.

Doing it that way would avoid this new copy class entirely, and avoid all the getLandingPage changes too.

**Review by williewillus — CHANGES_REQUESTED** [review]
Thanks for adding docs.

File: Common/src/main/java/vazkii/patchouli/client/book/gui/GuiBookPamphletLanding.java (hunk 3 above)
Comment (williewillus) [reply Common/src/main/java/vazkii/patchouli/client/book/gui/GuiBookPamphletLanding.java]: It would also avoid the need to have pamphlet_category entirely. Instead, you'd have something like pamphlet_mode: bool to enable this feature (or we could always enable it?)

File: Common/src/main/java/vazkii/patchouli/client/book/page/abstr/PageWithText.java [hunk 4]
```diff
@@ -11,7 +11,7 @@
 
 public abstract class PageWithText extends BookPage {
 
-	IVariable text;
+	protected IVariable text;
```
Comment (williewillus) [inline Common/src/main/java/vazkii/patchouli/client/book/page/abstr/PageWithText.java]: unused change now that the text stuff is moved to another diff

File: Common/src/main/java/vazkii/patchouli/client/book/page/PageText.java [hunk 5]
```diff
@@ -5,13 +5,18 @@
...
 import vazkii.patchouli.client.book.page.abstr.PageWithText;
 
 public class PageText extends PageWithText {
 
 	String title;
 
+	public void setText(String text) {
+		this.text = IVariable.wrap(text);
+	}
+
```
Comment (williewillus) [inline Common/src/main/java/vazkii/patchouli/client/book/page/PageText.java]: same

**Review by williewillus — COMMENTED** [review]
Looks good now, thanks

One last nit

**Comment by gamma-delta** [comment]
fiiiiiiiiiiiiiiiiiiiiine i did your thing :P
